api/src/services/schema.ts

```diff
@@ -0,0 +1,290 @@
+import { AbstractServiceOptions, DatabaseClients, Snapshot, SnapshotDiff, SnapshotWithHash } from '../types';
+import { applyDiff } from '../utils/apply-diff';
+import { getSnapshot, getVersionedHash } from '../utils/get-snapshot';
+import { getSnapshotDiff } from '../utils/get-snapshot-diff';
+
+const snapshotJoiSchema = Joi.object({
+	version: Joi.number().required(),
+	directus: Joi.string().required(),
+	vendor: Joi.string()
+		.valid(...DatabaseClients)
+		.optional(),
+	collections: Joi.array().items(
+		Joi.object({
+			collection: Joi.string(),
+			meta: Joi.any(),
+			schema: Joi.object({
+				name: Joi.string(),
+			}),
+		})
+	),
+	fields: Joi.array().items(
+		Joi.object({
+			collection: Joi.string(),
+			field: Joi.string(),
+			meta: Joi.any(),
+			schema: Joi.object({
+				default_value: Joi.any(),
+				max_length: [Joi.number(), Joi.string(), Joi.valid(null)],
+				is_nullable: Joi.bool(),
+			})
+				.unknown()
+				.allow(null),
+			type: Joi.string()
+				.valid(...TYPES, ...ALIAS_TYPES)
+				.allow(null),
+		})
+	),
+	relations: Joi.array().items(
+		Joi.object({
+			collection: Joi.string(),
+			field: Joi.string(),
+			meta: Joi.any(),
+			related_collection: Joi.any(),
+			schema: Joi.any(),
+		})
+	),
+});
+
+const deepDiffSchema = Joi.object({
+	kind: Joi.string().valid('N', 'D', 'E', 'A').required(),
+	path: Joi.array().items(Joi.string()),
+	lhs: Joi.any().when('kind', { is: 'N', then: Joi.optional(), otherwise: Joi.any() }),
+	rhs: Joi.any().when('kind', { is: 'D', then: Joi.optional(), otherwise: Joi.any() }),
+	index: Joi.any().when('kind', { is: 'A', then: Joi.number(), otherwise: Joi.optional() }),
+	item: Joi.any().when('kind', { is: 'A', then: Joi.any(), otherwise: Joi.optional() }),
+});
+
+const applyJoiSchema = Joi.object({
+	hash: Joi.string().required(),
+	diff: Joi.object({
+		collections: Joi.array().items(
+			Joi.object({
+				hash: Joi.string().when('diff.0.kind', { is: 'N', then: Joi.optional(), otherwise: Joi.any() }),
+				collection: Joi.string(),
+				diff: Joi.array().items(deepDiffSchema),
+			})
+		),
+		fields: Joi.array().items(
+			Joi.object({
+				hash: Joi.string().when('diff.0.kind', { is: 'N', then: Joi.optional(), otherwise: Joi.any() }),
+				collection: Joi.string(),
+				field: Joi.string(),
+				diff: Joi.array().items(deepDiffSchema),
+			})
+		),
+		relations: Joi.array().items(
+			Joi.object({
+				hash: Joi.string().when('diff.0.kind', { is: 'N', then: Joi.optional(), otherwise: Joi.any() }),
+				collection: Joi.string(),
+				field: Joi.string(),
+				related_collection: Joi.any(),
+				diff: Joi.array().items(deepDiffSchema),
+			})
+		),
+	}).required(),
+});
+
+export class SchemaService {
+	knex: Knex;
+	accountability: Accountability | null;
+
+	constructor(options: Omit<AbstractServiceOptions, 'schema'>) {
+		this.knex = options.knex || getDatabase();
+		this.accountability = options.accountability || null;
```

Comment:
Lets use `??` for defaults like this to avoid accidental falsey accidents

Reply:
Updated!

These lines are technically referred from other existing services, so I assume we should update them as well? Perhaps another PR?

Reply:
> These lines are technically referred from other existing services, so I assume we should update them as well? Perhaps another PR?

That's right! Yes, let's do it in another PR this one is already big 😄🎉